### PR TITLE
Fit Multiple Peaks

### PR DESCRIPTION
**Pull request by searscr**
# Overview
This PR creates a algorithm/recipe/service to call fit peaks for each spectrum in a workspace and plot the result to be observed by the user. This will serve as part of a larger workflow to check instrument calibration. 

## Testing
A button has been added to the test panel to perform a preconfigured peak fitting and plotting to show functionality.
Appropriate Unit Tests have also been included to be ran by `pytest`

### CODE REVIEW

File: src/snapred/ui/widget/WorkflowNode.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def startWorkflow(action, subview):
+    return WorkflowNodeModel(action=action, view=subview, nextModel=None)
+
+
+def continueWorkflow(workflow, action, subview):
+    currentWorkflow = workflow
+    while currentWorkflow.nextModel is not None:
+        currentWorkflow = workflow.nextModel
```
Comment (rboston628) [inline src/snapred/ui/widget/WorkflowNode.py:13]: This doesn't look like it does the correct thing.  This looks like it'd be an infinite loop, unless there's something else happening here

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    from snapred.backend.dao.calibration.Calibration import Calibration
+    from snapred.backend.dao.CrystallographicInfo import CrystallographicInfo
+    from snapred.backend.dao.FitMultiplePeaksIngredients import FitMultiplePeaksIngredients
+    from snapred.backend.recipe.algorithm.FitMultiplePeaksAlgorithm import (
+        FitMultiplePeaksAlgorithm,  # noqa: E402
+    )
+    from snapred.meta.Config import Resource
+
+    def test_init():
+        """Test ability to initialize purge overlapping peaks algo"""
```
Comment (rboston628) [inline tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py:21]: update comments

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        instrumentState = Calibration.parse_raw(
+            Resource.read("/inputs/purge_peaks/input_parameters.json")
```
Comment (rboston628) [inline tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py:23]: update resources? or explain

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        wsName = "testWS"
+        fitIngredients = FitMultiplePeaksIngredients(
+            InstrumentState=instrumentState, CrystalInfo=crystalInfo, InputWorkspace=wsName
+        )
+        fmpAlgo = FitMultiplePeaksAlgorithm()
+        fmpAlgo.initialize()
+        fmpAlgo.setProperty("FitMultiplePeaksIngredients", fitIngredients.json())
+        fmpAlgo.execute()
+        wsGroup = fmpAlgo.getProperty("OutputWorkspaceGroup").value
+        assert wsGroup == "fitPeaksWSGroup"
```
Comment (rboston628) [inline tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py:51]: Can we add better testing, such as ensuring proper values inside wsGroup?  At minimum, check what values *are* and assert those, so we can check future edits produce same output.

File: tests/unit/backend/recipe/algorithm/test_PurgeOverlappingPeaksAlgorithm.py [hunk 5]
```diff
@@ -22,35 +21,26 @@ def test_init():
...
-            Resource.read("/inputs/purge_peaks/input_ingredients.json")
-        ).reductionState.stateConfig.focusGroups
         crystalInfo = CrystallographicInfo.parse_raw(Resource.read("/inputs/purge_peaks/input_crystalInfo.json"))
         purgeAlgo = PurgeOverlappingPeaksAlgorithm()
         purgeAlgo.initialize()
         purgeAlgo.setProperty("InstrumentState", instrumentState.json())
-        purgeAlgo.setProperty("FocusGroups", json.dumps([focusGroup.json() for focusGroup in focusGroups]))
         purgeAlgo.setProperty("CrystalInfo", crystalInfo.json())
         assert purgeAlgo.getProperty("InstrumentState").value == instrumentState.json()
-        assert purgeAlgo.getProperty("FocusGroups").value == json.dumps(
```
Comment (rboston628) [inline tests/unit/backend/recipe/algorithm/test_PurgeOverlappingPeaksAlgorithm.py:35]: Can we assert the number of groups is what it should be here?

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                FitWindowBoundaryList=",".join(np.array(peakLimits).astype("str")),
+                OutputWorkspace=f"{wsName}_fitted_peakpositions_{index}",
+                OutputPeakParametersWorkspace=f"{wsName}_fitted_params_{index}",
+                BackgroundType="Quadratic",
+                FittedPeaksWorkspace=f"{wsName}_fitted_{index}",
+                ConstrainPeakPositions=True,
+                OutputParameterFitErrorsWorkspace=f"{wsName}_fitted_params_err_{index}",
+            )
+            self.mantidSnapper.executeQueue()
+            ws_group.add(f"{wsName}_fitted_peakpositions_{index}")
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:84]: Can these names be declared as constants somewhere earlier?

File: src/snapred/backend/dao/FitMultiplePeaksIngredients.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from snapred.backend.dao.CrystallographicInfo import CrystallographicInfo
+from snapred.backend.dao.state.InstrumentState import InstrumentState
+
+
+class FitMultiplePeaksIngredients(BaseModel):
+    """Class to hold fit multiple peaks parameters"""
+
+    InstrumentState: InstrumentState
+    CrystalInfo: CrystallographicInfo
+    PeakType: Optional[str] = "Gaussian"
```
[the remark this answers is not in the document]
Comment (searscr) [reply src/snapred/backend/dao/FitMultiplePeaksIngredients.py]: I misunderstood the use of `Optional`, it has been removed.

File: src/snapred/backend/recipe/algorithm/PurgeOverlappingPeaksAlgorithm.py [hunk 8]
```diff
@@ -37,17 +34,18 @@ def PyExec(self):
         multiplicity = np.array(crystalInfo.multiplicities)
         dSpacing = np.array(crystalInfo.dSpacing)
         A = fSquared * multiplicity * dSpacing**4
-        minimumA = np.max(A) * 0.05
+        maxA = np.max(A) * 0.05
```
[the remark this answers is not in the document]
Comment (searscr) [reply src/snapred/backend/recipe/algorithm/PurgeOverlappingPeaksAlgorithm.py]: Thank you for the clarification, it has been changed to better reflect what is happening.

File: src/snapred/backend/recipe/algorithm/PurgeOverlappingPeaksAlgorithm.py [hunk 9]
```diff
@@ -17,14 +14,14 @@ class PurgeOverlappingPeaksAlgorithm(PythonAlgorithm):
...
-        self.declareProperty("FocusGroups", defaultValue="", direction=Direction.Input)
         self.declareProperty("CrystalInfo", defaultValue="", direction=Direction.Input)
+        self.declareProperty("NumFocusGroups", defaultValue="", direction=Direction.Input)
         self.declareProperty("OutputPeakMap", defaultValue="", direction=Direction.Output)
         self.setRethrows(True)
 
     def PyExec(self):
         instrumentState = InstrumentState.parse_raw(self.getProperty("InstrumentState").value)
-        focusGroups = parse_raw_as(List[FocusGroup], self.getProperty("FocusGroups").value)
+        numFocusGroups = int(self.getProperty("NumFocusGroups").value)
```
[the remark this answers is not in the document]
Comment (searscr) [reply src/snapred/backend/recipe/algorithm/PurgeOverlappingPeaksAlgorithm.py:24]: The algorithm was only ever using the positional index of the `FocusGroup` to get information from the `InstrumentState` object. The contents of the FocusGroup were never used so I changed it to use the index only.

File: src/snapred/ui/widget/WorkflowNode.py (hunk 1 above)
Comment (searscr) [reply src/snapred/ui/widget/WorkflowNode.py:13]: Michael wrote this as part of the approval UI workflow, I'm not sure how works when executed.

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py (hunk 2 above)
Comment (searscr) [reply tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py]: Thank you, comments have been updated to reflect the correct algorithm name

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py (hunk 3 above)
Comment (searscr) [reply tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py:23]: Fit Multiple Peaks calls Purge Overlapping Peaks using that information so the resources used for testing are the same.

File: tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py (hunk 4 above)
Comment (searscr) [reply tests/unit/backend/recipe/algorithm/test_FitMultiplePeaksAlgorithm.py]: I don't want to test the values of the resulting workspaces because then I am testing result of Mantid `FitPeaks` and there is already extensive testing done in mantid to ensure that algorithm is working. Also I get different resulting values depending on which version of Mantid I am using. I have updated the test to check that the list of workspaces created in the workspace group are the correct workspaces. That should be the same every time no matter the result of `FitPeaks`

File: tests/unit/backend/recipe/algorithm/test_PurgeOverlappingPeaksAlgorithm.py (hunk 5 above)
Comment (searscr) [reply tests/unit/backend/recipe/algorithm/test_PurgeOverlappingPeaksAlgorithm.py:35]: This has been removed from the algorithm, it no longer uses Focus Groups

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py (hunk 6 above)
Comment (searscr) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:84]: These names are the names of the output workspaces created by `FitPeaks` the workspace name is provided by the user and the index corresponds to the number of histograms in the workspace. I could create a list of ws names prior to calling `FitPeaks` but I don't see how that would be beneficial?

File: src/snapred/ui/widget/WorkflowNode.py (hunk 1 above)
Comment (rboston628) [reply src/snapred/ui/widget/WorkflowNode.py:13]: @walshmm

Comment (walshmm) [reply src/snapred/ui/widget/WorkflowNode.py:13]: @rboston628 `nextModel` is `None` by default, this is just iterating a linked list of workflow objects

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py (hunk 6 above)
Comment (walshmm) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:84]: To Reece's point, you can declare the template strings earlier and format them for the inputs here.

some psuedoish code:
```
workspaceNameTemplate = "{wsName}_fitted_params_err_{index}"
...
    OutputParameterFitErrorsWorkspace=workspaceNameTemplate.format(wsName=wsName, index=index)",
)
```

Comment (rboston628) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:84]: That's what I meant, since you use each name twice.
```
ws_fittted_peaks = f"{wsName}_fitted_peakpositions_{index}"
ws_fitted_params = f"{wsName}_fitted_params_{index}"
ws_fitted = f"{wsName}_fitted_{index}"
ws_fitted_params_err = f"{wsName}_fitted_params_err_{index}"
 
 FitWindowBoundaryList=",".join(np.array(peakLimits).astype("str")),
                OutputWorkspace=ws_fitted_peaks,
                OutputPeakParametersWorkspace=ws_fitted_params,
                BackgroundType="Quadratic",
                FittedPeaksWorkspace=ws_fitted,
                ConstrainPeakPositions=True,
                OutputParameterFitErrorsWorkspace=ws_fitted_params_err,
            )
 ws_group.add(ws_fitted_peaks)
 ws_group.add(ws_fitted_params)
 ws_group.add(ws_fitted)
 ws_group.add(ws_fitted_params_err)
```

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        instrumentState = fitPeakIngredients.InstrumentState
+        crystalInfo = fitPeakIngredients.CrystalInfo
+        peakType = fitPeakIngredients.PeakType
+
+        beta_0 = instrumentState.gsasParameters.beta[0]
+        beta_1 = instrumentState.gsasParameters.beta[1]
+        FWHMMultiplierLeft = instrumentState.fwhmMultiplierLimit.minimum
+        FWHMMultiplierRight = instrumentState.fwhmMultiplierLimit.maximum
+        L = instrumentState.instrumentConfig.L1 + instrumentState.instrumentConfig.L2
+        ws = mtd[wsName]
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:37]: `ws = self.mantidSnapper.mtd[wsName]`
should work with latest

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+                ConstrainPeakPositions=True,
+                OutputParameterFitErrorsWorkspace=fittedParamsErr,
+            )
+            self.mantidSnapper.executeQueue()
+            ws_group.add(fittedPeakPos)
+            ws_group.add(fittedParams)
+            ws_group.add(fittedWS)
+            ws_group.add(fittedParamsErr)
+
+        DeleteWorkspace("ws2fit")
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py:93]:
```
self.mantidSnapper.DeleteWorkspace(
    "Deleting fitting workspace...",
    "ws2fit",
)
```
and can be moved to above L87

Comment (searscr) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py]: This gets overwritten with each iteration of the loop so I only need to delete it at the end of the loop. I made the change to use `mantidSnapper`

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
+import json
+
+import numpy as np
+from mantid.api import AlgorithmFactory, PythonAlgorithm, WorkspaceGroup, mtd
+from mantid.kernel import Direction
+from mantid.simpleapi import DeleteWorkspace
```
[the remark this answers is not in the document]
Comment (searscr) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py]: Unfortunately I still need to have `mtd` and `WorkspaceGroup`. I tested replacing `mtd` and `WorkspaceGroup` with corresponding `mantidSnapper` calls and I get errors. I don't think this is possible yet...

File: src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py (hunk 10 above)
Comment (searscr) [reply src/snapred/backend/recipe/algorithm/FitMultiplePeaksAlgorithm.py]: Thanks, I made the requested change.

**Review by rboston628 — APPROVED** [review]
